cpp/src/decisiontree/decisiontree.cuh

```diff
@@ -230,127 +227,81 @@ tl::Tree<T, T> build_treelite_tree(const DT::TreeMetaDataNode<T, L>& rf_tree,
-  int n_unique_labels = -1;  // number of unique labels in dataset
-  double prepare_time = 0;
-  double train_time   = 0;
-  MLCommon::TimerCPU prepare_fit_timer;
-  DecisionTreeParams tree_params;
-
  public:
-  /**
-   * @brief Fits a DecisionTree on given input data and labels
-   * @param[in] handle             cuML handle
-   * @param[in] data               pointer to input training data
-   * @param[in] ncols              number of features (columns)
-   * @param[in] nrows              number of samples (rows)
-   * @param[in] labels             pointer to label data
-   * @param[in] rowids             pointer to array of row indices mapping to data
-   * @param[in] n_sampled_rows     count of rows sampled
-   * @param[in] unique_labels      count of unique labels
-   * @param[in] is_classifier      true if task is classification, else false
-   * @param[in,out] tree           pointer to tree structure
-   * @param[in] tree_parameters    structure of tree parameters
-   * @param[in] seed               random seed
-   * @param[in] d_global_quantiles device pointer to global quantiles
-   */
-  void fit(const raft::handle_t& handle,
-           const T* data,
-           const int ncols,
-           const int nrows,
-           const L* labels,
-           unsigned int* rowids,
-           const int n_sampled_rows,
-           int unique_labels,
-           DT::TreeMetaDataNode<T, L>*& tree,
-           DecisionTreeParams tree_parameters,
-           uint64_t seed,
-           T* d_global_quantiles)
+  template <class DataT, class LabelT>
+  static std::shared_ptr<DT::TreeMetaDataNode<DataT, LabelT>> fit(
+    const raft::handle_t& handle,
+    const DataT* data,
+    const int ncols,
+    const int nrows,
+    const LabelT* labels,
+    MLCommon::device_buffer<std::size_t>* sampled_rows,
+    int unique_labels,
+    DecisionTreeParams params,
+    uint64_t seed,
+    std::shared_ptr<MLCommon::device_buffer<DataT>> quantiles,
+    int treeid)
   {
-    this->tree_params = tree_parameters;
-    this->prepare_fit_timer.reset();
-    const char* CRITERION_NAME[] = {"GINI", "ENTROPY", "MSE", "MAE", "END"};
-    CRITERION default_criterion =
-      (std::numeric_limits<L>::is_integer) ? CRITERION::GINI : CRITERION::MSE;
-    CRITERION last_criterion =
-      (std::numeric_limits<L>::is_integer) ? CRITERION::ENTROPY : CRITERION::MSE;
-
-    validity_check(tree_params);
-
-    if (tree_params.split_criterion ==
+    if (params.split_criterion ==
         CRITERION::CRITERION_END) {  // Set default to GINI (classification) or MSE (regression)
-      tree_params.split_criterion = default_criterion;
+      CRITERION default_criterion =
+        (std::numeric_limits<LabelT>::is_integer) ? CRITERION::GINI : CRITERION::MSE;
+      params.split_criterion = default_criterion;
+    }
+    // Dispatch objective
+    if (params.split_criterion == CRITERION::GINI) {
+      return Builder<GiniObjectiveFunction<DataT, LabelT>>(handle,
```

Comment:
By removing the `grow_tree`, we would lose pure C interface, but I am not sure C interface is really important. May be one of the @dantegd @JohnZed @teju85 can comment on need of C interface.